In [2]:
import numpy as np
import pandas as pd

In [3]:
insurance= pd.read_csv("/content/insurance_data.csv")
employee = pd.read_csv("/content/employee_data.csv")
vendor = pd.read_csv("/content/vendor_data.csv")

# TASK 1

In [4]:
insurance.shape

(10000, 38)

In [5]:
employee.shape

(1200, 10)

In [6]:
vendor.shape

(600, 7)

In [8]:
#Merging insurance and employee dataset throught AGENT_ID
df_1 = pd.merge(left= insurance,right=employee,on="AGENT_ID",how= "left")

In [9]:
#Merging df_1 and vendor dataset through VENDOR_ID
df = pd.merge(left= df_1,right=vendor,on="VENDOR_ID",how= "left")



# TASK **2**

In [10]:
#TOP 3 INSURANCE TYPE
df["INSURANCE_TYPE"].value_counts()[:3]


Property    1692
Mobile      1692
Health      1690
Name: INSURANCE_TYPE, dtype: int64

# TASK 3

In [12]:
# TOP 5 STATES with most insurance claims
df[df["RISK_SEGMENTATION"]=='H'].STATE.value_counts()[:5]


CA    112
AR     85
CO     78
CT     77
KY     75
Name: STATE, dtype: int64

# TASk 4

In [14]:
# ADDING COLUMN with condition given in ques
df["COLOCATION"]=  ((df['STATE_x']== df['INCIDENT_STATE']) & (df['INCIDENT_STATE']==df['STATE_y'])).map({True:1 ,False:0})
df

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER  \
0     2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468   
1     2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594   
2     2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969   
3     2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292   
4     2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204   
...                   ...            ...         ...           ...   
9995  2021-06-30 00:00:00    TXN00009996   A00006029   PLC00007245   
9996  2021-06-30 00:00:00    TXN00009997   A00003555   PLC00002590   
9997  2021-06-30 00:00:00    TXN00009998   A00004230   PLC00006613   
9998  2021-06-30 00:00:00    TXN00009999   A00006694   PLC00001781   
9999  2021-06-30 00:00:00    TXN00010000   A00000390   PLC00005197   

     POLICY_EFF_DT     LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  \
0       2015-06-23  2020-05-16  2020-05-21         Health          157.13   
1       2018-04-21  2020-05-13  2020-05-18       Property          141.71   
2       2019-10-03  2020-05-21  2020-05-26       Property          157.24   
3       2016-11-29  2020-05-14  2020-05-19         Health          172.87   
4       2011-12-26  2020-05-17  2020-05-22         Travel           88.53   
...            ...         ...         ...            ...             ...   
9995    2014-04-11  2021-06-25  2021-06-26          Motor          105.75   
9996    2015-08-25  2021-06-12  2021-06-17           Life           64.59   
9997    2018-10-11  2021-06-18  2021-06-23         Health          183.50   
9998    2016-01-29  2021-06-16  2021-06-21          Motor          113.46   
9999    2014-09-17  2021-06-21  2021-06-24         Mobile            8.22   

      CLAIM_AMOUNT  ... POSTAL_CODE_y EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0             9000  ...         37217           35441155  LRZO53254369804934   
1            26000  ...         85301           44055451  XBYG14382904542874   
2            13000  ...         36105           55770246  WRXH14597428306333   
3            16000  ...         73071           87652976  YEPN16077902472499   
4             3000  ...         21061           76119019  HZBK86544578578547   
...            ...  ...           ...                ...                 ...   
9995          1000  ...         72703           35413882  LFOJ76068590051807   
9996         67000  ...         31405           91206790  AHEF60771774635030   
9997          8000  ...         40272          115241097  WREH18805592939657   
9998          2000  ...         72703           13995380  AAZI21883736448048   
9999           300  ...         73003           39790996  RPES05613584687957   

                           VENDOR_NAME              ADDRESS_LINE1  \
0                           Ramsey Inc       3114 West 20th Court   
1     Morales, Campbell and Washington      42 West Louise Street   
2                          Johnson Ltd    2619 North Quality Lane   
3                     Blackwell-Foster          3609 Meadow Drive   
4                          Evans-Scott  4001 19th Place Northeast   
...                                ...                        ...   
9995                               NaN                        NaN   
9996          Newton, Hodges and Henry      32512 Christine Drive   
9997                    Reese-Mckinney         20 Garfield Street   
9998       Harmon, Manning and Bradley           1823 Dinger Road   
9999                    Foster-Sanchez        4255 Tanager Common   

      ADDRESS_LINE2          CITY STATE  POSTAL_CODE  COLOCATION  
0               NaN   Panama City    FL      32405.0           0  
1                 B  Fayetteville    AR      72701.0           0  
2              #315  Fayetteville    AR      72703.0           0  
3               NaN     Nashville    TN      37215.0           0  
4               NaN    Washington    DC      20018.0           0  
...             ...           ...   ...          ...  

In [16]:
# Mean of COLOCATION
df["COLOCATION"].mean()

0.0044

# TASK 5


In [18]:
#UPDATING the AUTHORITY_CONTACTED column
df.loc[(df["POLICE_REPORT_AVAILABLE"] == 1)&(df['AUTHORITY_CONTACTED']!='Police'),"AUTHORITY_CONTACTED"]="Police"
df


TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER  \
0     2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468   
1     2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594   
2     2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969   
3     2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292   
4     2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204   
...                   ...            ...         ...           ...   
9995  2021-06-30 00:00:00    TXN00009996   A00006029   PLC00007245   
9996  2021-06-30 00:00:00    TXN00009997   A00003555   PLC00002590   
9997  2021-06-30 00:00:00    TXN00009998   A00004230   PLC00006613   
9998  2021-06-30 00:00:00    TXN00009999   A00006694   PLC00001781   
9999  2021-06-30 00:00:00    TXN00010000   A00000390   PLC00005197   

     POLICY_EFF_DT     LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  \
0       2015-06-23  2020-05-16  2020-05-21         Health          157.13   
1       2018-04-21  2020-05-13  2020-05-18       Property          141.71   
2       2019-10-03  2020-05-21  2020-05-26       Property          157.24   
3       2016-11-29  2020-05-14  2020-05-19         Health          172.87   
4       2011-12-26  2020-05-17  2020-05-22         Travel           88.53   
...            ...         ...         ...            ...             ...   
9995    2014-04-11  2021-06-25  2021-06-26          Motor          105.75   
9996    2015-08-25  2021-06-12  2021-06-17           Life           64.59   
9997    2018-10-11  2021-06-18  2021-06-23         Health          183.50   
9998    2016-01-29  2021-06-16  2021-06-21          Motor          113.46   
9999    2014-09-17  2021-06-21  2021-06-24         Mobile            8.22   

      CLAIM_AMOUNT  ... POSTAL_CODE_y EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0             9000  ...         37217           35441155  LRZO53254369804934   
1            26000  ...         85301           44055451  XBYG14382904542874   
2            13000  ...         36105           55770246  WRXH14597428306333   
3            16000  ...         73071           87652976  YEPN16077902472499   
4             3000  ...         21061           76119019  HZBK86544578578547   
...            ...  ...           ...                ...                 ...   
9995          1000  ...         72703           35413882  LFOJ76068590051807   
9996         67000  ...         31405           91206790  AHEF60771774635030   
9997          8000  ...         40272          115241097  WREH18805592939657   
9998          2000  ...         72703           13995380  AAZI21883736448048   
9999           300  ...         73003           39790996  RPES05613584687957   

                           VENDOR_NAME              ADDRESS_LINE1  \
0                           Ramsey Inc       3114 West 20th Court   
1     Morales, Campbell and Washington      42 West Louise Street   
2                          Johnson Ltd    2619 North Quality Lane   
3                     Blackwell-Foster          3609 Meadow Drive   
4                          Evans-Scott  4001 19th Place Northeast   
...                                ...                        ...   
9995                               NaN                        NaN   
9996          Newton, Hodges and Henry      32512 Christine Drive   
9997                    Reese-Mckinney         20 Garfield Street   
9998       Harmon, Manning and Bradley           1823 Dinger Road   
9999                    Foster-Sanchez        4255 Tanager Common   

      ADDRESS_LINE2          CITY STATE  POSTAL_CODE  COLOCATION  
0               NaN   Panama City    FL      32405.0           0  
1                 B  Fayetteville    AR      72701.0           0  
2              #315  Fayetteville    AR      72703.0           0  
3               NaN     Nashville    TN      37215.0           0  
4               NaN    Washington    DC      20018.0           0  
...             ...           ...   ...          ...  

# TASK 6

In [19]:
# MAX date in dataset
df.TXN_DATE_TIME.max()

'2021-06-30 00:00:00'

In [20]:
#Current date is the max date of the dataset
C_Date = pd.to_datetime('2021-06-30')

In [21]:
# converting column to datetime
df['TXN_DATE_TIME'] = pd.to_datetime(df['TXN_DATE_TIME'])

In [22]:
# data of last 30 days
Last_30_Days = df.loc[(df['TXN_DATE_TIME'] > '2021-05-30') & (df['TXN_DATE_TIME'] < '2021-06-30')] 

In [23]:
# AVERAGE CLaim of last 30 days
AVERAGE_CLAIM = Last_30_Days.groupby('INSURANCE_TYPE').mean()['CLAIM_AMOUNT'].to_dict()

In [24]:
#calculating claim deviation
def Claim_deviation(row):
    if (C_Date - row['TXN_DATE_TIME'])<pd.Timedelta(days=30):
        return -1
    ClAIM_DEVIATION = AVERAGE_CLAIM[row['INSURANCE_TYPE']] / row['CLAIM_AMOUNT']
    if ClAIM_DEVIATION < 0.5:
        return 1
    else:
        return 0

In [25]:

df['CLAIM_DEVIATION'] = df.apply(Claim_deviation,axis=1)

In [26]:
df

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0       2020-06-01    TXN00000001   A00003822   PLC00008468    2015-06-23   
1       2020-06-01    TXN00000002   A00008149   PLC00009594    2018-04-21   
2       2020-06-01    TXN00000003   A00003172   PLC00007969    2019-10-03   
3       2020-06-01    TXN00000004   A00007572   PLC00009292    2016-11-29   
4       2020-06-01    TXN00000005   A00008173   PLC00000204    2011-12-26   
...            ...            ...         ...           ...           ...   
9995    2021-06-30    TXN00009996   A00006029   PLC00007245    2014-04-11   
9996    2021-06-30    TXN00009997   A00003555   PLC00002590    2015-08-25   
9997    2021-06-30    TXN00009998   A00004230   PLC00006613    2018-10-11   
9998    2021-06-30    TXN00009999   A00006694   PLC00001781    2016-01-29   
9999    2021-06-30    TXN00010000   A00000390   PLC00005197    2014-09-17   

         LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  \
0     2020-05-16  2020-05-21         Health          157.13          9000   
1     2020-05-13  2020-05-18       Property          141.71         26000   
2     2020-05-21  2020-05-26       Property          157.24         13000   
3     2020-05-14  2020-05-19         Health          172.87         16000   
4     2020-05-17  2020-05-22         Travel           88.53          3000   
...          ...         ...            ...             ...           ...   
9995  2021-06-25  2021-06-26          Motor          105.75          1000   
9996  2021-06-12  2021-06-17           Life           64.59         67000   
9997  2021-06-18  2021-06-23         Health          183.50          8000   
9998  2021-06-16  2021-06-21          Motor          113.46          2000   
9999  2021-06-21  2021-06-24         Mobile            8.22           300   

      ... EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0     ...           35441155  LRZO53254369804934   
1     ...           44055451  XBYG14382904542874   
2     ...           55770246  WRXH14597428306333   
3     ...           87652976  YEPN16077902472499   
4     ...           76119019  HZBK86544578578547   
...   ...                ...                 ...   
9995  ...           35413882  LFOJ76068590051807   
9996  ...           91206790  AHEF60771774635030   
9997  ...          115241097  WREH18805592939657   
9998  ...           13995380  AAZI21883736448048   
9999  ...           39790996  RPES05613584687957   

                           VENDOR_NAME              ADDRESS_LINE1  \
0                           Ramsey Inc       3114 West 20th Court   
1     Morales, Campbell and Washington      42 West Louise Street   
2                          Johnson Ltd    2619 North Quality Lane   
3                     Blackwell-Foster          3609 Meadow Drive   
4                          Evans-Scott  4001 19th Place Northeast   
...                                ...                        ...   
9995                               NaN                        NaN   
9996          Newton, Hodges and Henry      32512 Christine Drive   
9997                    Reese-Mckinney         20 Garfield Street   
9998       Harmon, Manning and Bradley           1823 Dinger Road   
9999                    Foster-Sanchez        4255 Tanager Common   

     ADDRESS_LINE2          CITY STATE POSTAL_CODE  COLOCATION  \
0              NaN   Panama City    FL     32405.0           0   
1                B  Fayetteville    AR     72701.0           0   
2             #315  Fayetteville    AR     72703.0           0   
3              NaN     Nashville    TN     37215.0           0   
4              NaN    Washington    DC     20018.0           0   
...            ...           ...   ...         ...         ...   
9995           NaN           NaN   NaN         NaN           0   
9996           NaN    Union City    CA     94587.0           0   
9997           NaN       Needham    MA      2492.0           0   
9998           NaN     Annapolis    MD   

In [27]:
df.CLAIM_DEVIATION.value_counts()

 0    9216
-1     784
Name: CLAIM_DEVIATION, dtype: int64

9216 claims deviation are less than 0.5, 
784 claims transaction have  history of less than 30 days

# TASK 7

In [30]:
# INSURANCE type
Type_Count = df.groupby('AGENT_ID')['INSURANCE_TYPE'].nunique()

#AGENTS that have more than 2 types
Agent_Insurance_Types = Type_Count>2

#Agents worked in 2 or more
Multi_Type_Agents = Agent_Insurance_Types[Agent_Insurance_Types].index

#TOP AGENTS
TOP_AGENTS = df.groupby(['AGENT_ID','AGENT_NAME']).sum().loc[Multi_Type_Agents][['CLAIM_AMOUNT']].sort_values('CLAIM_AMOUNT',ascending=False)

In [31]:
TOP_AGENTS

CLAIM_AMOUNT
AGENT_ID   AGENT_NAME                    
AGENT00807 Don Filkins             528800
AGENT00679 Clara Barnett           489000
AGENT00771 Roger Burns             422100
AGENT00125 Salvador Soriano        400400
AGENT00789 Alison Baron            392900
...                                   ...
AGENT00910 Vincent Ragan            15500
AGENT00689 Connie Hunt              13700
AGENT00732 Sylvia Tran              11300
AGENT00706 Laura Staggs             10800
AGENT00604 Thelma Salinas            5500

[1164 rows x 1 columns]

In [32]:
TOP_AGENTS.join(Type_Count)

CLAIM_AMOUNT  INSURANCE_TYPE
AGENT_ID   AGENT_NAME                                    
AGENT00807 Don Filkins             528800               6
AGENT00679 Clara Barnett           489000               4
AGENT00771 Roger Burns             422100               5
AGENT00125 Salvador Soriano        400400               4
AGENT00789 Alison Baron            392900               4
...                                   ...             ...
AGENT00910 Vincent Ragan            15500               3
AGENT00689 Connie Hunt              13700               3
AGENT00732 Sylvia Tran              11300               3
AGENT00706 Laura Staggs             10800               3
AGENT00604 Thelma Salinas            5500               3

[1164 rows x 2 columns]

# TASK 8

In [41]:
Premium_Collected_Earlier = df['PREMIUM_AMOUNT'].sum()
Premium_Collected_Earlier

885085.95

In [34]:
#total premium collected
Prem_By_Types = df.groupby('INSURANCE_TYPE').sum()['PREMIUM_AMOUNT'].to_dict()
Prem_By_Types

{'Health': 252455.73,
 'Life': 125621.22,
 'Mobile': 15191.42,
 'Motor': 165391.98,
 'Property': 202285.83,
 'Travel': 124139.77}

In [35]:
# Mobile & Travel premium discounted by 10%
Prem_By_Types['Mobile']*=0.9
Prem_By_Types['Travel']*=0.9
# Health and Property premium increases by 7%
Prem_By_Types['Health']*=1.07
Prem_By_Types['Property']*=1.07
# Life and Motor premium increased by 2%
Prem_By_Types['Life']*=1.02
Prem_By_Types['Motor']*=1.02

In [38]:
# Premium after increase and discount
New_Premium = sum(Prem_By_Types.values())
New_Premium

908805.0042000001

In [39]:
(New_Premium-Premium_Collected_Earlier)*100/Premium_Collected_Earlier

2.6798588543858517

Overall change in Premium amount collected is  2.679% or 2.68%

# TASK 9

In [43]:
#calculating discount for given condition
Discount=(df['TENURE']>60)&(df['EMPLOYMENT_STATUS']=='N')&(df['NO_OF_FAMILY_MEMBERS']>=4)

In [44]:
df['ELIGIBLE_FOR_DISCOUNT'] = Discount.astype('int')

In [45]:
df['ELIGIBLE_FOR_DISCOUNT'].mean()

0.0299

Total 2.99% customers are eligible for discount

# TASK 10

In [46]:
C_Date

Timestamp('2021-06-30 00:00:00')

In [53]:
#data for last 3 days
Last_3_Days = df.loc[(df['TXN_DATE_TIME'] > C_Date-pd.Timedelta(days=4)) & (df['TXN_DATE_TIME'] < C_Date)]

In [54]:
#total approved claims in last 30 days
Claims_In_Last_30_Days = pd.DataFrame.from_dict(Last_30_Days[Last_30_Days['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST30DAYS'])

In [55]:
#total approved claims in last 3 days
Claims_In_Last_3_Days = pd.DataFrame.from_dict(Last_3_Days[Last_3_Days['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['NO_OF_CLAIMS_IN_LAST3DAYS'])

In [56]:
#total approved claims
Approved_Claims = Claims_In_Last_30_Days.join(Claims_In_Last_3_Days)

In [57]:
Approved_Claims['CLAIM_VELOCITY'] = Approved_Claims['NO_OF_CLAIMS_IN_LAST30DAYS']/Approved_Claims['NO_OF_CLAIMS_IN_LAST3DAYS']

In [58]:
Approved_Claims

NO_OF_CLAIMS_IN_LAST30DAYS  NO_OF_CLAIMS_IN_LAST3DAYS  \
Health                           128                         14   
Life                             128                         12   
Mobile                           127                          9   
Motor                            117                         12   
Property                         125                         12   
Travel                           123                         14   

          CLAIM_VELOCITY  
Health          9.142857  
Life           10.666667  
Mobile         14.111111  
Motor           9.750000  
Property       10.416667  
Travel          8.785714

# TASK 11

In [59]:
Lowest_5 = df.groupby('AGENT_ID').sum().quantile(0.05)['CLAIM_AMOUNT']

In [60]:
Agent_Total_Data = df.groupby('AGENT_ID').sum()

In [63]:
#Low performing agent data
Low_Performing_Agents = Agent_Total_Data[Agent_Total_Data['CLAIM_AMOUNT'] <= Lowest_5][['CLAIM_AMOUNT']]

In [64]:
Low_Performing_Agents

CLAIM_AMOUNT
AGENT_ID                
AGENT00006         22400
AGENT00014          7400
AGENT00016         27000
AGENT00085         25700
AGENT00088         27700
AGENT00091         30000
AGENT00098         28200
AGENT00134         22000
AGENT00149         28900
AGENT00156         13200
AGENT00172         22200
AGENT00185         16100
AGENT00186         27700
AGENT00188         28400
AGENT00197         18500
AGENT00252         20000
AGENT00263         11500
AGENT00275         16200
AGENT00279         28000
AGENT00300         12600
AGENT00310         28800
AGENT00316         22200
AGENT00351         21700
AGENT00425         17800
AGENT00442         30300
AGENT00536         20000
AGENT00562         21100
AGENT00571         24200
AGENT00585         16600
AGENT00590         22200
AGENT00604          5500
AGENT00621          5300
AGENT00628          6100
AGENT00672         21300
AGENT00689         13700
AGENT00706         10800
AGENT00721          9000
AGENT00732         11300
AGENT00773         15000
AGENT00780         11300
AGENT00790          6000
AGENT00804         20000
AGENT00805         22600
AGENT00830         23500
AGENT00841         17800
AGENT00856         24200
AGENT00894         27600
AGENT00910         15500
AGENT00969         28300
AGENT00985         30000
AGENT01021         24300
AGENT01037          5600
AGENT01052         28000
AGENT01077          4200
AGENT01107         24700
AGENT01128         28000
AGENT01131         27500
AGENT01151         28000
AGENT01154           800
AGENT01179         27300

# TASK 12

In [65]:
Condition = (df['CLAIM_STATUS']=='A')&(df['RISK_SEGMENTATION']=='H')&(df['INCIDENT_SEVERITY']=='Major Loss')

In [66]:
Criteria= df[Condition]

In [67]:
Suspicious_Data = Criteria.groupby('AGENT_ID').sum()

In [68]:
Suspicious_Agents = list(Suspicious_Data[Suspicious_Data['CLAIM_AMOUNT']>=15000].index)

In [69]:
employee['SUSPICIOUS'] = 0

In [70]:
result = employee['AGENT_ID'].isin(Suspicious_Agents)

In [71]:
employee.loc[result,'SUSPICIOUS']=1

In [72]:
employee['SUSPICIOUS'].mean()

0.12416666666666666

This means from all enployee 12.41% are suspicious